### PR TITLE
Move C API versioning into Version.xcconfig

### PR DESCRIPTION
**Pull request by dmbryson**

### CODE REVIEW

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: products/libllbuild/include/llbuild/llbuild.h [hunk 1]
```diff
@@ -47,7 +47,9 @@
 /// 1: Added `environment` parameter to llb_buildsystem_invocation_t.
 ///
 /// 0: Pre-history
```
Comment (neonichu) [inline products/libllbuild/include/llbuild/llbuild.h]: I think we should probably move this comment over to the xcconfig as well.

File: products/llbuildSwift/BuildSystemBindings.swift [hunk 2]
```diff
@@ -18,6 +18,10 @@ import Foundation
 
 import llbuild
 
+#if !LLBUILD_C_API_VERSION_4
+  import llbuild_c_api_version_unsupported
```
Comment (aciidb0mb3r) [inline products/llbuildSwift/BuildSystemBindings.swift]: We can now use `#error("Unsupported API version")`

Comment (dmbryson) [reply products/llbuildSwift/BuildSystemBindings.swift]: Doesn’t that require Swift 5 (based on the PR I found earlier)?

Sounds good to me, it wasn’t clear if it was going to be available in needed targets.

File: products/libllbuild/include/llbuild/llbuild.h (hunk 1 above)
Comment (dmbryson) [reply products/libllbuild/include/llbuild/llbuild.h]: I wasn’t sure the best place. The xcconfig seems logical, but isn’t installed with the headers.  We are shipping the xcconfig with the framework, though, so maybe its fine.

File: products/llbuildSwift/BuildSystemBindings.swift (hunk 2 above)
Comment (aciidb0mb3r) [reply products/llbuildSwift/BuildSystemBindings.swift]: It looks like it was implemented for 4.2: https://github.com/apple/swift-evolution/blob/master/proposals/0196-diagnostic-directives.md

We usually track Swift master but you can maintain compatibility using:

```swift
#if swift(>=4.2)
  #error("Unsupported API version")
#else
  import llbuild_c_api_version_unsupported
#endif
```

File: products/libllbuild/include/llbuild/llbuild.h (hunk 1 above)
Comment (ddunbar) [reply products/libllbuild/include/llbuild/llbuild.h]: Hmmm... tough question. Let's leave the authoritative source as the code (easily found by jump-to-definition), and reference this comment in the xcconfig.

File: Xcode/Configs/Version.xcconfig [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// Define the C API version
+LLBUILD_C_API_VERSION = 4
+
+LLBUILD_VERSION_DEFINITIONS = LLBUILD_C_API_VERSION=$(LLBUILD_C_API_VERSION) LLBUILD_C_API_VERSION_$(LLBUILD_C_API_VERSION)
```
Comment (ddunbar) [inline Xcode/Configs/Version.xcconfig]: Probably good to have a comment here explaining why we have both?

File: products/libllbuild/CMakeLists.txt [hunk 4]
```diff
@@ -22,15 +22,30 @@ set_property(TARGET libllbuild PROPERTY MACOSX_RPATH ON)
 
 target_link_libraries(libllbuild ${DEPENDENCIES})
 
+# Load version number from the XCConfig
+file(STRINGS "${CMAKE_SOURCE_DIR}/Xcode/Configs/Version.xcconfig" VERSION_XCCONFIG)
+string(REGEX REPLACE ".*LLBUILD_C_API_VERSION = ([0-9]+).*" "\\1" LLBUILD_C_API_VERSION "${VERSION_XCCONFIG}")
+
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/include/llbuild/version.h.in ${CMAKE_CURRENT_BINARY_DIR}/include/llbuild/version.h)
```
Comment (ddunbar) [inline products/libllbuild/CMakeLists.txt]: Nice, thanks!

File: products/libllbuild/CMakeLists.txt [hunk 5]
```diff
@@ -22,15 +22,30 @@ set_property(TARGET libllbuild PROPERTY MACOSX_RPATH ON)
 
 target_link_libraries(libllbuild ${DEPENDENCIES})
 
+# Load version number from the XCConfig
+file(STRINGS "${CMAKE_SOURCE_DIR}/Xcode/Configs/Version.xcconfig" VERSION_XCCONFIG)
+string(REGEX REPLACE ".*LLBUILD_C_API_VERSION = ([0-9]+).*" "\\1" LLBUILD_C_API_VERSION "${VERSION_XCCONFIG}")
```
Comment (ddunbar) [inline products/libllbuild/CMakeLists.txt]: Can we see if we can make CMake error out of this regex fails to match properly? If its non-trivial we can skip it.

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Pull request merged**